diesel/src/pg/types/numeric.rs

```diff
@@ -0,0 +1,113 @@
+            let mut decimal_part = &absolute - absolute.with_scale(0);
+            // scale is the amount of digits to print. to_string() includes a "0.",
+            // that's why the -2 is there.
+            let scale = decimal_part.to_string().len() as u16 - 2;
+            let mut weight = 0;
+
+            // Encode the integer part
+            while ten_k < integer_part {
+                weight += 1;
+                // digit is integer_part REM 10_000
+                let (div, digit) = integer_part.div_rem(&ten_k);
+                digits.push(digit.to_u16().expect("digit < 10000, but cannot fit in i16") as i16);
+                integer_part = div;
+            }
+            digits.push(integer_part.to_string().parse::<i16>().expect("digit < 10000, but cannot fit in i16"));
+
+            digits.reverse();
+
+            // Encode the decimal part
+            let ten_k = BigDecimal::from_str("10000").unwrap();
+            while decimal_part != BigDecimal::zero() {
+                decimal_part *= &ten_k;
+                let digit = decimal_part.to_bigint().expect("Can always take integer part of BigDecimal");
+                // This can be simplified when github.com/akubera/bigdecimal-rs/issues/13 gets
+                // solved; decimal_part -= &digit; should suffice by then.
+                decimal_part -= BigDecimal::new(digit.clone(), 0);
+                digits.push(digit.to_u16().expect("digit < 10000, but cannot fit in i16") as i16);
+            }
+
+            let numeric = match self.sign() {
+                Sign::Plus => {
+                    PgNumeric::Positive {
+                        digits, scale, weight
+                    }
+                }
+                Sign::Minus => {
+                    PgNumeric::Negative {
+                        digits, scale, weight
+                    }
+                }
+                Sign::NoSign => {
+                    PgNumeric::Positive {
+                        digits: vec![0],
+                        scale: 0,
+                        weight: 0,
+                    }
+                }
+            };
+            ToSql::<types::Numeric, Pg>::to_sql(&numeric, out)
+        }
+    }
+
+    impl FromSql<types::Numeric, Pg> for BigDecimal {
+        fn from_sql(numeric: Option<&[u8]>) -> Result<Self, Box<Error+Send+Sync>> {
+            let numeric = try!(PgNumeric::from_sql(numeric));
+            let (sign, weight, _, digits) = match numeric {
+                PgNumeric::Positive {
+                    weight, scale, digits
+                } => {
+                    (Sign::Plus, weight, scale, digits)
+                },
+                PgNumeric::Negative {
+                    weight, scale, digits
+                } => {
+                    (Sign::Minus, weight, scale, digits)
+                },
+                PgNumeric::NaN => {
+                    return Err(Box::from("NaN"));
```

Comment:
If nothing else, this error message should more specifically state that the reason this is treated as an error is that the bigdecimal crate does not provide support for `NaN`, not that Diesel wishes to treat it as an error